Language specific Multimedia for Modules and Forms

sravfeyn opened the pull request:
Spec and discussion http://manage.dimagi.com/default.asp?159795#929276
`suite.xml` spec https://confluence.dimagi.com/display/ccinternal/Proposed+suite.xml+changes+for+localizable+menu+multimedia

sravfeyn committed:
Language specific media for Forms/modules - data model change

sravfeyn committed:
Display language specific media in Form/Module edit page

sravfeyn committed:
handle edits correctly for lang-specific form/module media

sravfeyn committed:
handle Bulk App Translations to use lang-specific media for forms/modules

sravfeyn commented:
Test are going to fail for now.

sravfeyn committed:
wrap old app-docs for lang-specific media

sravfeyn committed:
manage icon/audio edits in once place and enforce convention

sravfeyn committed:
add app-strings for localized media

sravfeyn committed:
change bulk-app translation xlsx files for localized media

sravfeyn committed:
Add basic localized media test

sravfeyn committed:
first attempt at suite Spec for localized media

sravfeyn committed:
util methods

sravfeyn committed:
merge master, resovle conflict

sravfeyn committed:
make media-suite actually work

sravfeyn committed:
Merge branch 'master' into langMedia

sravfeyn committed:
fix textordisplay

sravfeyn committed:
a better localized class, refactor

sravfeyn committed:
wrap correctly

sravfeyn committed:
media localization strings for case-list form

sravfeyn committed:
refactor into seperate classes

sravfeyn committed:
generate correct localized suite

sravfeyn committed:
comment

sravfeyn commented:
@snopoke  Hey Simon, this is now in a good state for review. Will add some more tests, fix old-data and write the migration script.

@orangejenny (buddy)

sravfeyn committed:
add some media tests

sravfeyn committed:
fix errors in localized suite

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,71 @@ def validate(self, value, required=True):
+        Properties are map of lang-code to filepath
+
+        Convention (for old docs): Unless a specific language key exists,
+        by default, media for that language is same as that for English
```
Is this literally English, or the app's default language?

orangejenny replied:
Similarly, I don't quite follow the places where this mixin uses 'en' - what happens if the app doesn't use english?

sravfeyn replied:
I planned to deal with this almost at the end after getting everything in place. This is what I am going to do.
Since this class doesn't have access to app's default-language, I will add a special key called 'default' instead of 'en'. The behaviour will be as following
- All migrated old-docs will have this `default` key pointing to the media that was already set
- If a particular lang key doesn't exist the `default` media is returned. 
- One can set a lang-specific media, by editing icon/image while viewing app in that language
- If app's default language changes, the media pointed by `default` will be changed to that.

sravfeyn committed:
fix old tests

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,71 @@ def validate(self, value, required=True):
+            old_media = getattr(self, media_attr)
+            if old_media is None or isinstance(old_media, (str, unicode)):
+                # ToDo - better way than to hardcode default-language.
+                new_media = {'en': old_media} if old_media else {}
```
Similarly to above, what happens here if the app doesn't use english at all?

orangejenny commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,71 @@ def validate(self, value, required=True):
+        self = super(NavMenuItemMediaMixin, cls).wrap(data)
+
+        # ToDo - Remove after migration
+        for media_attr in ('media_image', 'media_audio'):
```
Would there be value to making these constants instead of strings?

sravfeyn replied:
I could get that list by accessing attributes on cls, but if someone adds an extra attribute latter this would break, so I wrote it in a tuple. Also this code will be removed after the initial migration.

sravfeyn committed:
change default behaviour for localized media

sravfeyn committed:
change old test source files

sravfeyn committed:
fix bug and change test

sravfeyn commented on corehq/apps/app_manager/tests/data/bulk_app_translation/download/app.json:
```diff
@@ -1 +1,447 @@
-{"build_langs": ["en"], "auto_gps_capture": false, "deployment_date": null, "user_type": null, "text_input": "roman", "secure_submissions": false, "build_broken": false, "show_user_registration": false, "phone_model": null, "build_signed": true, "recipients": "", "copy_history": [], "platform": "nokia/s40", "application_version": "2.0", "is_released": false, "admin_password": null, "build_spec":  ...
+{
```
This change is similar to https://github.com/dimagi/commcare-hq/pull/6503/files#diff-d9bc9308147e25e5268d7141d167306dR280
Plus one line to many lines.

sravfeyn committed:
add case-list-form suite test and fix bugs. refactor tests

sravfeyn committed:
Add no media test

sravfeyn commented on corehq/apps/app_manager/id_strings.py:
```diff
@@ -202,6 +202,37 @@ def media_resource(multimedia_id, name):
+    )
+
+
+# ToDo - figure out whether media-strings are custom or not
```
hey @snopoke what are 'non "app-string" id strings'? or should all of these be decorated with `pattern` making them custom app strings?

snopoke replied:
Yea, these shouldn't be included in `get_app_translations` so they should have a pattern decorator (I think)

sravfeyn commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1425,12 +1522,24 @@ def entries(self):
-                    media_image=form.media_image,
-                    media_audio=form.media_audio,
-                )
+                # ToDo: remove version-hardcode
```
Is there any better way to do version check?

snopoke replied:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/feature_support.py#L5-5

snopoke replied:
Comment no longer needed

sravfeyn committed:
set case-label when creating a module

sravfeyn committed:
Add some more tests

sravfeyn committed:
remove todos

sravfeyn committed:
don't wrap if none

sravfeyn committed:
wrap correctly

sravfeyn committed:
exclude empty paths

sravfeyn committed:
basic Migration experiment

sravfeyn commented:
hey @snopoke this is good for review now. Will remove `hold-off-merge` tag tomorrow once I figure out migration.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,91 @@ def validate(self, value, required=True):
+    @staticmethod
+    def _check_media_attribute(media_attr):
+        if media_attr not in ('media_image', 'media_audio'):
+            raise Exception("Unknown media attribute %s." % media_attr)
```
this could just be an assert

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,91 @@ def validate(self, value, required=True):
+
+        media_dict = getattr(self, media_attr) or {}
+        media_dict[lang] = media_path or ''
+        setattr(self, media_attr, media_dict)
```
What the purpose of `media_path or ''`. When will `media_path` be `None`?

Also you can just go `self[media_attr][lang] = media_path or ''`

sravfeyn replied:
Only if the caller passes `None`

sravfeyn replied:
Didn't know can treat `self` as a `dict`!

snopoke replied:
Jsonobject makes sure it isn't None

sravfeyn replied:
I got the `None` `TypeError` when I tried that, because migrated apps will have default as `None` instead of `{}`

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,91 @@ def validate(self, value, required=True):
+    def _all_media_paths(self, media_attr):
+        self._check_media_attribute(media_attr)
+        media_dict = getattr(self, media_attr) or {}
+        valid_media_paths = [media for media in media_dict.values() if media]
```
Won't this contain duplicates because of the `default` key?

snopoke commented on corehq/apps/app_manager/tests/test_media_suite.py:
```diff
@@ -1,24 +1,135 @@
+
 
-class MediaSuiteTest(SimpleTestCase, TestFileMixin):
+class MediaSuiteTest(TestCase, TestFileMixin):
```
Why did the change to this test require DB access?

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -226,11 +226,73 @@ def set_display(self, locale_id=None, media_image=None, media_audio=None):
+        if media_text:
+            self.display = LocalizedMediaDisplay(
+                text=text,
+                media_text=[text] + media_text if text else [] + media_text
```
no need for the empty list at the end: `media_text=[text] + media_text if text else media_text`

sravfeyn replied:
Haha...the way I wrote it

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -226,11 +226,73 @@ def set_display(self, locale_id=None, media_image=None, media_audio=None):
+
+        if media_text:
+            self.display = LocalizedMediaDisplay(
+                text=text,
```
`LocalizedMediaDisplay` doesn't seem to have a `text` property

snopoke commented on corehq/apps/app_manager/app_strings.py:
```diff
@@ -83,6 +85,8 @@ def maybe_add_index(text):
+            yield id_strings.form_icon_locale(form), form.icon_by_language(lang)
+            yield id_strings.form_audio_locale(form), form.audio_by_language(lang)
 
         if hasattr(module, 'case_list_form') and module.case_list_form.form_id:
```
Presumably you also need to add the ones for case list form media?

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,91 @@ def validate(self, value, required=True):
 
 
 class NavMenuItemMediaMixin(DocumentSchema):
+    """
```
I'm not sure about this whole `default` thing. I'm wondering if we shouldn't do it in the same way we do for things like `form.name` which is just to return *any* item if the one you are looking for isn't there:
```python
langs = request.GET['lang'] + app.langs
item.get_icon(langs)

...

def get_icon(langs):
    for lang in langs:
        if lang in media_dict and media_dict[lang]:
            return media_dict[lang]
    for lang, media in sorted(media_dict.items()):
        return media
```
See:
* https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/views.py#L624-624
* https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/templatetags/xforms_extras.py#L18-18

Let me know if you want to go to voice to chat about this.

sravfeyn replied:
This was the initial direction I took, I used 'en' by default. But I went with this solution as this gives more unambiguous. For example an app that doesn't have 'en' as one of the languages will have 'default' key in this approach instead of 'en'. Apart from that, all of the 'default' logic is automatically taken care of in the interfacing methods.

snopoke replied:
I think its still fine to use `default` in the migration and in `wrap` but after that I don't think its necessary to keep it updated if you use the same pattern as we do for form name translations.

sravfeyn replied:
Will chat with you on Skype

snopoke commented:
Generally looks good. Just the question about the 'default' key approach.

snopoke commented:
Also build fail is real

snopoke commented:
======================================================================
ERROR: test_report_module (corehq.apps.app_manager.tests.test_suite.SuiteTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/tests/test_suite.py", line 743, in test_report_module
    app.create_app_strings('default'),
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/models.py", line 3886, in create_app_strings
    return gen.create_default_app_strings(self)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/app_strings.py", line 129, in create_default_app_strings
    self.create_app_strings(app, lc, for_default=True)
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/app_manager/app_strings.py", line 117, in create_app_strings
    return commcare_translations.dumps(messages).encode('utf-8')
  File "submodules/commcare-translations/commcare_translations.py", line 51, in dumps
    if not val.strip():
AttributeError: 'NoneType' object has no attribute 'strip'

sravfeyn committed:
Merge branch 'master' into langMedia

sravfeyn committed:
don't generate locale if there's no media

sravfeyn committed:
respond to code review

sravfeyn committed:
remove wrong file addition

sravfeyn committed:
keep correct migration

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,90 @@ def validate(self, value, required=True):
+    def _all_media_paths(self, media_attr):
+        self._check_media_attribute(media_attr)
+        media_dict = getattr(self, media_attr) or {}
+        valid_media_paths = set([media for media in media_dict.values() if media])
```
Not a biggy but you can use set comprehension `{m for m in list if m}`

sravfeyn committed:
code review

sravfeyn committed:
remove default key convention, follow form-name convention

sravfeyn commented:
hey @snopoke I removed the default key stuff, do you think this is good now? I will put this on staging for QA, if so.

snopoke commented on corehq/apps/app_manager/models.py:
```diff
@@ -986,9 +986,80 @@ def validate(self, value, required=True):
+
+    @property
+    def default_media_audio(self):
+        return self.audio_by_language('en')
```
A comment for these methods would be good explaining that they are for legacy apps.

snopoke commented:
I think this is good to go for QA. Thanks

sravfeyn committed:
comment

sravfeyn commented:
Thanks Simon. Sent out QA ticket http://manage.dimagi.com/default.asp?168748

millerdev committed:
Merge remote-tracking branch 'origin/fix-time-ago-sort' into langMedia+fix-time-ago-sort

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into langMedia

esoergel committed:
Make _get_item_media more failure tolerant

@sravfeyn You might wanna take a look at this.  This was failing on
staging because another branch added a new reference to this function
which didn't update the `to_language`.  It looks like you might also
just be able to do this here anyways

esoergel committed:
Merge branch 'langMedia' of github.com:dimagi/commcare-hq into langMedia+fix-time-ago-sort

esoergel committed:
Merge branch 'fix-time-ago-sort' of github.com:dimagi/commcare-hq into langMedia+fix-time-ago-sort

Conflicts:
	corehq/apps/app_manager/translations.py

TylerSheffels committed:
Merge remote-tracking branch 'origin/langMedia' into case-list-menu-icons+langMedia

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py
	corehq/apps/app_manager/translations.py

TylerSheffels committed:
Add localization for case list menu icons

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into langMedia

esoergel committed:
Merge branch 'langMedia' of github.com:dimagi/commcare-hq into case-list-menu-icons+langMedia

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

TylerSheffels committed:
Merge remote-tracking branch 'origin/master' into case-list-menu-icons+langMedia

Conflicts:
	corehq/apps/app_manager/tests/test_media_suite.py

dimagimon commented on corehq/apps/app_manager/tests/test_media_suite.py:
```diff
@@ -45,3 +53,106 @@ def test_update_image_id(self):
         self.assertEqual(old_image.unique_id, new_image.unique_id)
         self.assertNotEqual(old_image.version, new_image.version)
+
+class LocalizedMediaSuiteTest(TestCase, TestFileMixin):
```
E302 expected 2 blank lines, found 1

proteusvacuum committed:
Merge branch 'master' into langMedia

emord commented:
@sravfeyn could you merge master into this? Rebuilt staging and noticed there's a merge conflict on binary xlsx files (and it looks like the build is failing anyways so I cant tell if I'm doing the right thing). Thanks!

sravfeyn commented:
Sorry about that will do.

sravfeyn committed:
Merge branch 'master' into langMedia

sravfeyn commented:
https://github.com/dimagi/commcare-hq/pull/7041
